doc/metadata-server/README.md

```diff
@@ -0,0 +1,327 @@
+
+### Terminology
+
+- Marlowe contract (`Contract`) - a particular contract expression which can be executed on-chain which has all participants and values (like amounts, timeouts etc.) set.
+
+- Marlowe contract transaction (`MarloweTx`)- transaction on the block chain which is a part of the ongoing Marlowe contract.
+
+- Marlowe contract metadata (`ContractMetadata`) - a particular set of information which we want to attach to a Marlowe contract (description, documents / agreements etc.)
+
+- Role distribution (`RoleDistribution`) - when contract uses roles as `Party` values we use NFTs to represent roles on-chain. The initial transaction pay these NFTs to some addresses so it defines this initial distribution.
+
+- Transaction metadata (`TxMetadata`) - on-chain information ([`CBOR` encoded](https://github.com/input-output-hk/cardano-node/blob/master/doc/reference/tx-metadata.md)) which is attached to the transaction . It is inaccessible to the transaction validators and minting policy scripts.
+
+- Marlowe contract transaction metadata (`MarloweTxMetadata`) - metadata attached to the on-chain transaction executing Marlowe Contract (`MarloweTx`) related to the `Contract`.
+
+
+### `Contract` extraction
+
+When initiator submits a `Contract` to the blockchain it builds up a transaction which outputs define role tokens distribution. Additionally transaction contains "continuation" output which is "guarded" by `Marlowe` interpreter script. Role tokens outputs are assigned to a set of addresses chosen by the `Tx` author. We can interpret them as invitations to the contract or permissions to be the contract party (this tokens shows up in the user wallets).
+
+Because continuation output from the initial `Tx` contains the full code ("modulo" `Merkleization`) of the `Contract` we are able to [extract it](https://github.com/input-output-hk/marlowe-cardano/blob/f452d98942a14d34c10f2963ec4b84147da5621c/marlowe/src/Language/Marlowe/Client/History.hs). In other words when someone is "invited" to the already initiated on-chain contract we are able to extract the Marlowe code from that transaction and present it to the possible participant for an audit, simulation etc. before he / she commits to it. We can do this at any stage of the contract progression because we can track the chain back to the initial transaction and perform the extraction.
+
+<!-- [ IMAGE -> Extraction of the contract -> Simulation / Audit ] -->
+
+### `MarloweTxMetadata`
+
+The submitter of the transaction have an ability to attach additional [`CBOR` encoded data to it](https://github.com/input-output-hk/cardano-node/blob/master/doc/reference/tx-metadata.md). This data are not accessible to the transaction validators so they can be arbitrary information not necessarily valid and related to the `Tx` content. To simplify the discussion for now and assume that there are no size limits and costs associated with `TxMetadata` and we use this layer to store all contract metadata.
```

Comment:
- Change "have an ability" to "has the ability".
- Change "This data are" to "These data are" or "This data is".